### ISSUE TITLE
Middle-ground betwixt `ast` and `libcst` for parsing source code, and emitting only what changed in CST from AST

### ISSUE THREAD

**Issue by SamuelMarks** [Instagram/LibCST#564]
What I'm imagining is using `ast` to find and modify nodes, and when ready to emit (unparse) align with the CST so that only what the AST has changed will be modified.

Some caveats:
- Cannot delete nodes (I assume)
- Maybe difficulty in converting single line to multiline, and multiline to single line

What do you think?

> I am modifying AST nodes, changing: docstrings; `ast.Assign`; `ast.AnnAssign`; and `ast.FunctionDef`/`ast.AsyncFunctionDef`. Inferring types, adding them as a type comment xor annotation; converting between docstring formats (incl. adding/remove types); and updating the `return` attribute of a function definition.

Currently my library only uses [`ast`](https://docs.python.org/3/library/ast.html) and [`inspect`](https://docs.python.org/3/library/inspect.html) for this… which means it can support 3.6, 3.7, 3.8, 3.9, 3.10 & 3.11. https://github.com/SamuelMarks/cdd-python

…but adding/removing type annotations, for example, has the [unexpected] side-effect of removing all comments and reducing whitespace.

Tossing up betwixt rewriting my library around LibCST | redbaron and hacking together a basic parser & emitter to replace `ast.parse` and `ast.unparse` with some line-aware copypasta.

Thanks for sharing your perspective

**Comment by zsol** [comment]
I don't fully understand what you're proposing to do. If your intention is to have your library rewrite existing source code then you'll need something like LibCST and build your library on top of it.

**Comment by SamuelMarks** [comment]
@zsol I'm more thinking a multi-linked data-structure, from Python's `ast` to `libcst` (and back again). That way transformations can be done at a high level (AST) and the modifications can be done at a low level (CST: whitespace aware).

The attributes should be sufficient to make this link, assuming the changes don't delete nodes + the change has the same number of lines as the original:

```coffeescript
attributes (int lineno, int col_offset, int? end_lineno, int? end_col_offset)
```

**Comment by zsol** [comment]
Oh, I see! Interesting idea. Before going into questions about the implementation (I'm not exactly sure how changes from AST could be mapped to relevant changes at the libcst layer), let me ask this: why do you think the libcst API is not high-level enough? What type of transformation would be easier to do on an AST than a CSTNode? Do you have a concrete example at hand?

**Comment by SamuelMarks** [comment]
The main thing is my library: https://github.com/SamuelMarks/cdd-python

If I could get away with just rewriting my `emit.file`—plus some `line_no` & `col_offset` mapping—then the whole project could be reused in a whitespace-aware manner.

Switching to LibCST completely would require a more major refactor (and possibly even a rewrite). Also by relying on builtin libraries, I can support every maintained version of Python (incl. 3.11a3).

**Comment by zsol** [comment]
After thinking about this a bit, I don't think I want to commit to maintaining a mapping like this: a feature like this is inherently fragile considering the language is changing in non-trivial ways every year. I'm happy to be convinced otherwise, but I'd need to see at least a draft implementation that convinces me the maintenance burden is negligible.

**Comment by SamuelMarks** [comment]
@zsol I wrote my own CST and added it to that library of mine (link above).

My target is adding/removing/changing of: docstrings, function return types, function arguments, and Assign/AnnAssign. All but the last are now implemented.

I was careful not to replace code elsewhere in my codebase, so everything except my new CST code (in its own files) stays, and everything else works exclusively with the builtin [`ast`](https://docs.python.org/3/library/ast.html) module as before.
